Additional Build Documentation added.

SayantanMajumdar opened the pull request:
In response to #788 added file `docs/Additional_Build_Documentation.md` containing
additional build documentation which attempts to accieve the following:
   - demonstarte the use of BENCHMARK_DOWNLOAD_DEPENDENCIES
   - vcpkg/installation
   - Building from  head 
   - embedding Benchmark in other CMake projects

Then added a link to the file `docs/Additional_Build_Documentation.md` in `README.md`.

LebedevRI commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+
+Then to build Benchmark we use CMake again as follows:
+```bash
+$ cmake -build build
````
`--build` ?

SayantanMajumdar replied:
will fix it

LebedevRI commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+
+### In Linux
+```bash
+$ g++ -std=c++11 BenchmarkEverything.cpp -I<BENCHMARK_SOURCE_DIR>/include -pthread -L <BENCHMARK_BUILD_DIR>\src -lbenchmark
````
wrong slash

SayantanMajumdar replied:
will fix it

LebedevRI commented on docs/Additional_Build_Documentation.md:
```diff
@@ -0,0 +1,296 @@
+
+After building Benchmark we can choose to install it in our systems. 
+
+For Ubuntu and Debian based systems we have to build Benchmark in the Release configuration by applying `-DCMAKE_BUILD_TYPE=RELEASE` in the configuration step. Then install it using `sudo make install` inside the build directory.
```
please linewrap everything to 80 col

SayantanMajumdar replied:
what about the commands? do they have to be under 80 col?

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+```
+To configure Benchmark we use CMake as follows:
+```bash
+$ cmake -G <generator> [options] -S benchmark -B build
````
you should differentiate between the cmake and bazel steps.

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+
+Then to build Benchmark we use CMake again as follows:
+```bash
+$ cmake -build build
````
alternatively, users can use cmake to generate and then Visual Studio or whatever build system they want to build.

SayantanMajumdar replied:
yeah, will add that. I feared that might create a situation where we have to give instructions for each generator. but will add it.

SayantanMajumdar replied:
added example for VS 15 2017, the build documentation in `README.md` was good and so added the example there. and removed the build documentation form the file added by me.

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+```bash
+$ cmake -G <generator> [options] -S benchmark -B build
+```
+If we haven't cloned the googletest repository as described above we can ask benchmark to download this dependency in the above configuration step by adding `-DBENCHMARK_DOWNLOAD_DEPENDENCIES=ON`
````
this is only if you want to build from head. if you just want to _use_ benchmark then we should document that separately.

SayantanMajumdar replied:
using has been put in a separate file now

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+
+### In Windows
+```bash
+> cl /EHsc /MD BenchmarkEverything.cpp /I<BENCHMARK_SOURCE_DIR>\include /link shlwapi.lib <BENCHMARK_BUILD_DIR>\src\Release\benchmark.lib
````
i presume there's an easier way if you're using VS

SayantanMajumdar replied:
yes, others will have syntax similar to linux(I think)

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+
+### In Linux
+```bash
+$ g++ -std=c++11 BenchmarkEverything.cpp -I<BENCHMARK_SOURCE_DIR>/include -pthread -L <BENCHMARK_BUILD_DIR>\src -lbenchmark
````
if you're going to prescribe a compiler, consider clang++ :)

SayantanMajumdar replied:
done

dominichamon commented on README.md:
```diff
@@ -40,6 +40,7 @@ dependency can be provided two ways:
 If you do not wish to build and run the tests, add `-DBENCHMARK_ENABLE_GTEST_TESTS=OFF`
 to `CMAKE_ARGS`.
 
+For a more detailed documentation of building benchmark checkout [Additional Build Documentation](docs/Additional_Build_Documentation.md)
```
this also covers USING benchmark.

Perhaps the docs would be clearer if we separated these two use-cases. One doc for building and installing from source, one doc for using it (linking to building from source, but also having instructions for `apt` or `vcpkg` or whatever).

SayantanMajumdar replied:
made a separate file for embedding Benchmark and moved the `vcpkg` installation to `README.md`

dominichamon commented on docs/Additional_Build_Documentation.md:
````diff
@@ -0,0 +1,296 @@
+$ ./vcpkg install benchmark
+```
+
+# Embedding Benchmark in other projects
````
as above, this should come first _or_ be a different doc.

SayantanMajumdar replied:
embedding and using is in the file `docs/Using_Benchmark.md` and the installation using `vcpkg` has been moved to `README.md`

SayantanMajumdar commented:
Moved the installation documentation for `vcpkg` into the `README.md` and added a hint for the VS 15 2017 generator in the build instructions in `README.md`. I removed the build instructions in my docs file  `docs/Additional_Build_Documentation.md` as it was redundant and it was well written in the `README.md`.

Renamed the file `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` with instructions of using Benchmark and embedding it in other projects. 
Instructions on 
-    Using Benchmark with single file projects in the terminal
-    Including Benchmark as a submodule in a project.
-    Making CMake download and build Benchmark as a part of the project.

SayantanMajumdar commented:
Moved the installation instructions for `vcpkg` to `README.md` from `docs/Additional_Build_Documentation.md` and added a simple example for building Benchmark using the VS 15 2017 for VS users. The build instructions in `docs/Additional_Build_Documentation.md` were redundent and it was already well written in `README.md` .

Renamed `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` and included documentation for using Benchmark with small single file projects, documentation about using it as submodule in other projects and making CMake download and build Benchmark as a part of the project just like googletest

SayantanMajumdar commented:
Moved the installation instructions for `vcpkg` to `README.md` from `docs/Additional_Build_Documentation.md` and added a simple example for building Benchmark using the VS 15 2017 for VS users. The build instructions in `docs/Additional_Build_Documentation.md` were redundent and it was already well written in `README.md` .

Renamed `docs/Additional_Build_Documentation.md` to `docs/Using_Benchmark.md` and included documentation for using Benchmark with small single file projects, documentation about using it as submodule in other projects and making CMake download and build Benchmark as a part of the project just like googletest

LebedevRI commented on README.md:
````diff
@@ -29,6 +29,11 @@ $ cmake -G <generator> [options] ../benchmark
 
+Similarly for the Visual Studio 15 2017 generator we can use
+```bash
+> msbuild benchmark.sln /p:Configuration=Release /p:Platform=Win32
````
Do we want to spell it out for every cmake generator?
Maybe let's just
```diff
-$ make
+$ cmake --build ./
```
that is more correct than hardcoding `make` there anyway

LebedevRI commented on README.md:
````diff
@@ -70,6 +75,26 @@ If you need to install the library globally
+```
+For Linux
+```bash
+$ ./vcpkg install benchmark
````
These expect that `vcpkg` is located in current directory, and only differ in slashes.
Can this be condensed into a single entry?

SayantanMajumdar replied:
will fix it

LebedevRI commented on docs/Using_Benchmark.md:
```diff
@@ -0,0 +1,285 @@
+compiler which is provided with Visual Studio and also linked the `shlwapi` 
+library.
+
+## In Linux
```
This is super flame-wary and nitpicky, but i think linux should be explained first.

SayantanMajumdar replied:
its the same

LebedevRI commented on docs/Using_Benchmark.md:
```diff
@@ -0,0 +1,285 @@
+if(WIN32)
+    target_link_libraries(BenchmarkEverything shlwapi)
+elseif(LINUX)
+    target_link_libraries(BenchmarkEverything pthread)
```
This is a bad advise.
The library needs to be found first.

SayantanMajumdar replied:
I tried to keep it simple if the library can't be found the error will say that

LebedevRI replied:
link-time error
link-time errors are not nice :/

LebedevRI commented on docs/Using_Benchmark.md:
```diff
@@ -0,0 +1,285 @@
+to build, like `Release` or `Debug`. To run `BenchmarkEverything` we just 
+need to run the executable generated inside `<BUILD_DIR>`.
+
+## Use CMake for embedding Benchmark
```
There is one more variant - download it at cmake time, and then fall-back into `Adding Benchmark as a git submodule` variant.

SayantanMajumdar replied:
that would require you understand both given options first. someone doing that probably is an advanced coder
